feat(ffe-form-react): Make PhoneNumber accept FieldMessage

TobiasHansen opened the pull request:
BREAKING CHANGE: Replaces countryCodeInvalid and numberInvalid with countryCodeFieldMessage, numberFieldMessage and countryCodeAndNumberFieldMessage.

This commit changes how the components shows invalidation errors. It replaces countryCodeInvalid and numberInvalid with countryCodeFieldMessage, numberFieldMessage and countryCodeAndNumberFieldMessage. This makes it possible to render an error message instead of just making the input field red.

The design has been discussed with Kjersti Velten, Bernadette Ann Eirheim and Bjørg Andrea Lysbakken.

![image](https://user-images.githubusercontent.com/6777340/122347097-f7affc00-cf49-11eb-81ac-c2ce7cba1ee9.png)

Commit:
feat(ffe-form-react): Make PhoneNumber accept FieldMessage

BREAKING CHANGE: Replaces countryCodeInvalid and numberInvalid with countryCodeFieldMessage, numberFieldMessage and countryCodeAndNumberFieldMessage.

This commit changes how the components shows invalidation errors. It replaces countryCodeInvalid and numberInvalid with countryCodeFieldMessage, numberFieldMessage and countryCodeAndNumberFieldMessage. This makes it possible to render an error message instead of just making the input field red.

TobiasHansen commented on packages/ffe-form-react/src/PhoneNumber.js:
```diff
@@ -18,14 +20,42 @@ export default class PhoneNumber extends React.Component {
+            !!numberFieldMessage &&
+            !countryCodeAndNumberFieldMessage
+        ) {
+            throw new Error(
```
In this design if both countryCode and numberField is invalid you should be using the countryCodeAndNumberFieldMessage-prop. We found that other components in FFE used this way to throw errors if props are used wrong. For instance if you use both ToolTip and Description for the RadioButtonInputGroup-component.

Is there a better way to do this? As this would result in runtime errors, and might be hard to catch during development.

kwltrs replied:
Det blir litt mye å kaste en Error her. Ser ikke at det er såpass kritisk feil å få disse tre props feil og brukeropplevelsen av en systemfeil pga en caught eller uncaught Error veier tyngre enn en manglende feiltekst på en input.

TobiasHansen replied:
Yes, var i grunn det vi også tenkte, men prøvde å være konsekvent med de andre komponentene her, men da fjerner vi denne :+1:

TobiasHansen replied:
Vi la det inn som en console.error.

kwltrs commented on packages/ffe-form-react/src/PhoneNumber.js:
```diff
@@ -18,14 +20,42 @@ export default class PhoneNumber extends React.Component {
+            );
+        }
+
+        let fieldMessage = null;
```
Holder med  `let fieldMessage`, synes jeg.

kwltrs commented on packages/ffe-form-react/src/PhoneNumber.js:
```diff
@@ -18,14 +20,42 @@ export default class PhoneNumber extends React.Component {
+        const fieldMessageId = fieldMessage
+            ? (fieldMessage.props && fieldMessage.props.id) ||
+              this.fieldMessageId
+            : undefined;
```
Litt uheldig å definere en variable med `undefined`. Bedre å gjøre det samme prinsipp som på `fieldMessage`, dvs.

```js
let fieldMessagerId;
if (...) {
  fieldMessageId = ...
}
```

kwltrs commented on packages/ffe-form-react/src/PhoneNumber.js:
```diff
@@ -18,14 +20,42 @@ export default class PhoneNumber extends React.Component {
+        }
+
+        let fieldMessage = null;
+        if (!!countryCodeAndNumberFieldMessage) {
```
Trenger ikke denne cast til boolean med `!!`. Det er det som javascript runtime gjør allikevel i `if()`.

kwltrs commented on packages/ffe-form-react/src/PhoneNumber.js:
```diff
@@ -65,7 +103,16 @@ export default class PhoneNumber extends React.Component {
+                                aria-invalid={String(
+                                    !!countryCodeFieldMessage ||
+                                        !!countryCodeAndNumberFieldMessage,
+                                )}
```
Ikke nødvendig å caste boolean til string her. React gjør det for deg. Kanskje litt tydeligere hva som er greia her om man skriver det som `aria-invalid={!!(countryCodeFileMessage || countryCodeAndNumberFieldMessage)}`?

kwltrs commented on packages/ffe-form-react/src/PhoneNumber.js:
```diff
@@ -18,14 +20,42 @@ export default class PhoneNumber extends React.Component {
+              this.fieldMessageId
+            : undefined;
+
+        const hasFieldMessage = !!fieldMessage;
```
Synes denne variable bidrar ikke så mye til lesbarhet av koden.

kwltrs commented on packages/ffe-form/less/phone-number.less:
```diff
@@ -20,8 +20,6 @@
 
 .ffe-phone-number {
-    display: flex;
-
```
Jeg er litt usikker om denne endring burde være en egen commit. Henger den sammen med den breaking change i ffe-forms-react?
Jeg har aldri sett dobbel-scope i commit-meldingen, dvs de to pakkenavn i `feat(ffe-form, ffe-form-react)`. Fungerer det med conventional-commits?

TobiasHansen replied:
Det vet jeg egentlig ikke hvordan vil fungere i praksis, her henger disse veldig tett sammen for at stylingen skal bli riktig.

Så i praksis vil vi at dagens versjon av PhoneNumber skal bruke dagens versjon av ffe-form, mens den nye versjonen må også ha den nye versjonen av ffe-form. Hvordan gjøres dette best i praksis i designsystemet?

kwltrs replied:
Tror det blir best med én PR med to BREAKING CHANGE commits - en på ffe-from og en på ffe-form-react. Det er da så vi pleier å håndtere slik situasjoner i ffe-pakkene som du beskriver.

TobiasHansen replied:
Da er den splittet i to :)

kwltrs commented on packages/ffe-form-react/src/PhoneNumber.js:
```diff
@@ -18,14 +20,42 @@ export default class PhoneNumber extends React.Component {
+            numberFieldMessage,
+            countryCodeAndNumberFieldMessage,
             className,
+            extraMargin = true,
```
Den burde være i `defaultProps` litt lengre ned i filen. Kan bli forvirring å dele opp defaults på to sted.

kwltrs reviewed (commented):
Fant litt småpirk, men fin PR! Bra at du tenkte på testene og ts-typene.

andrea-sb1 commented:
Når det bare er feil på telefonnummer (ikke landskode) tenker jeg det er fint om feilmeldingen plasseres rett under telefonnummer-feltet (starter på linje med feltet). Vet ikke om det byr på noen tekniske utfordringer?

andrea-sb1 commented:
> Når det bare er feil på telefonnummer (ikke landskode) tenker jeg det er fint om feilmeldingen plasseres rett under telefonnummer-feltet (starter på linje med feltet). Vet ikke om det byr på noen tekniske utfordringer?

@TobiasHansen og jeg snakket om dette og landet på at vi beholder feilmeldingen der den er, under landskode. Det vil introdusere en del tekniske utfordringer å flytte meldingen kun ved feil på telefonnummer og vi er ikke sikre på at det vil gi bedre brukeropplevelse (det er bare en hypotese) :)

Commit:
feat(ffe-form): Update styling for PhoneNumber component

BREAKING CHANGE: Updates styling of PhoneNumber component after introducing fieldMessage props

kwltrs commented:
(Det er noe feil med travis-oppsettet. Jobber med saken)

kwltrs commented:
Må bare åpne og lukke denne PR for å la snyk resjekke status.
